pkg/operator/ceph/object/bucket/predicate.go

```diff
@@ -62,8 +64,19 @@ func predicateController(ctx context.Context, c client.Client) predicate.Funcs {
 						}
 					}
 				}
-			}
 
+			case *cephv1.CephObjectStore:
+				if objOld, ok := e.ObjectOld.(*cephv1.CephObjectStore); ok {
+					if objNew, ok := e.ObjectNew.(*cephv1.CephObjectStore); ok {
+						if objOld.Spec.IsExternal() && objNew.Spec.IsExternal() {
+							if objOld.Spec.Gateway.ExternalRgwEndpoints[0].IP != objNew.Spec.Gateway.ExternalRgwEndpoints[0].IP {
+								logger.Infof("there is an update event on external CephObjectStore CR spec endpoints %s %s, so reconcile the bucket provisioner", objOld.Spec.Gateway.ExternalRgwEndpoints[0].IP, objNew.Spec.Gateway.ExternalRgwEndpoints[0].IP)
+								return true
+							}
+						}
```

Comment:
We should consider the larger case here as well. I know that the bug we have is mentioned for external object stores, but I think this general bug could exist for internal object stores as well.

The generalized statement of the bug is, "when CephObjectStore endpoints change, OBCs are not updated." With that in mind, I think it may be better to check the `obj.Status.Endpoints` and return true if any of the endpoints change in the status. I think this also better reflects the real-world, where OBCs are consumers of the CephObjectStore. To keep provider-consumer relationships and APIs clear, it's best when consumers rely on the Status of the resource they rely on. If consumers rely on the Spec of the provider, then consumers may finish reconiling before the provider reconciliation finishes -- this is especially a problem if the new spec has an error that results in the provider being unable to finish reconciling.

---

Also, I am worried about bugs as a result of hard-coding comparison of endpoint `[0]`. In order to prevent bugs, I think it is better to make the predicate simple and return true in case there is any doubt. The doubts in my mind is this: the predicate does not control the behavior of the reconciler, and because the predicate doesn't reuse a method also used by the reconciler, predicate behavior can diverge easily in the future if something changes in the reconcile related to which endpoint in the list Rook uses for OBCs. There is therefore some long-term maintenance risk for the predicate to only focus on endpoint `[0]`.

I would prefer to err on the side of caution and risk Rook reconciling more often than necessary than risk Rook failing to reconcile when it should. In reality, these endpoints will rarely change, and there will rarely be more than 1, so the risk of reconciling more often should be very low.

Reply:
Added the resource compariter

Reply:
Updating to use internal mode too but 
the endpoint will always be `http://rook-ceph-rgw-my-store.rook-ceph.svc:8080` according to my understanding, the port may change